lib/connectors/transient_error_helper.rb

```diff
@@ -0,0 +1,43 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+#
+
+require 'utility/logger'
+require 'utility/exception_tracking'
+require 'utility/error_monitor'
+
+module Connectors
+  class TransientErrorHelper
+    def initialize(error_monitor)
+      @error_monitor = error_monitor
+    end
+
+    def yield_single_document(identifier: nil)
+      Utility::Logger.debug("Extracting single document for #{identifier}") if identifier
+      yield
+      @error_monitor.note_success
+    rescue *fatal_exception_classes => e
+      Utility::ExceptionTracking.augment_exception(e)
+      Utility::Logger.error("Encountered a fall-through error during extraction#{identifying_error_message(identifier)}: #{e.class}: #{e.message} {:message_id => #{e.id}}")
+      raise
+    rescue StandardError => e
+      Utility::ExceptionTracking.augment_exception(e)
+      Utility::Logger.warn("Encountered error during extraction#{identifying_error_message(identifier)}: #{e.class}: #{e.message} {:message_id => #{e.id}}")
+      @error_monitor.note_error(e, :id => e.id)
+    end
+
+    private
+
+    def identifying_error_message(identifier)
+      identifier.present? ? " of '#{identifier}'" : ''
+    end
+
+    def fatal_exception_classes
```

Comment:
This used to be overridable at the connector level. Now the user will need to subclass the error helper AND override the connector's `initialize` method in order to change it. What if instead, `fatal_exception_classes` was a method in `Base::Connector` that could be used as a parameter for this class?

Reply:
We can postpone refactoring cause only Mongo uses it. When we refactor (can do now too) there are 2 options:
1. inline transient_error_helper or whatever it's called into Base::Connector
2. Make users of connector create an instance of transient_error_helper and pass error classes into initializer.
3. Something else I could not think about?

Reply:
From Sean:
3. is a combo 1+2. Connector defines a function "fatal_exception_classes" and base connector calls it when initializing an instance of transient_error_helper.